phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/PhoenixMapReduceUtil.java

```diff
@@ -157,6 +192,216 @@ public static void setOutput(final Job job, final String tableName,final String
+
+                if (splitByStats) {
+                    for (Scan aScan : scans) {
+                        if (LOG.isDebugEnabled()) {
+                            LOG.debug("Split for  scan : " + aScan + "with scanAttribute : " + aScan
+                                    .getAttributesMap() + " [scanCache, cacheBlock, scanBatch] : ["
+                                    +
+                                    aScan.getCaching() + ", " + aScan.getCacheBlocks() + ", " + aScan.getBatch() + "] and  regionLocation : "
+                                    + regionLocation);
+                        }
+
+                        psplits.add(
+                                new PhoenixInputSplit(Collections.singletonList(aScan), regionSize,
+                                        regionLocation));
+                    }
+                } else {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Scan count[" + scans.size() + "] : " + Bytes
+                                .toStringBinary(scans.get(0).getStartRow()) + " ~ " + Bytes
+                                .toStringBinary(scans.get(scans.size() - 1).getStopRow()));
+                        LOG.debug("First scan : " + scans.get(0) + "with scanAttribute : " + scans
+                                .get(0).getAttributesMap() + " [scanCache, cacheBlock, scanBatch] : " +
+                                "[" + scans.get(0).getCaching() + ", " + scans.get(0).getCacheBlocks()
+                                + ", " + scans.get(0).getBatch() + "] and  regionLocation : " +
+                                regionLocation);
+
+                        for (int i = 0, limit = scans.size(); i < limit; i++) {
+                            LOG.debug("EXPECTED_UPPER_REGION_KEY[" + i + "] : " + Bytes
+                                    .toStringBinary(scans.get(i).getAttribute(
+                                            BaseScannerRegionObserver.EXPECTED_UPPER_REGION_KEY)));
+                        }
+                    }
+
+                    psplits.add(new PhoenixInputSplit(scans, regionSize, regionLocation));
+                }
+            }
+            return psplits;
+        }
+    }
+
+    public static ResultIterator initializeIterator(Configuration conf, List<Scan> scans,
+            QueryPlan queryPlan) throws SQLException, IOException {
+        List<PeekingResultIterator> iterators = Lists.newArrayListWithExpectedSize(scans.size());
+        StatementContext ctx = queryPlan.getContext();
+        ReadMetricQueue readMetrics = ctx.getReadMetricsQueue();
+        String tableName = queryPlan.getTableRef().getTable().getPhysicalName().getString();
+        String snapshotName = conf.get(PhoenixConfigurationUtil.SNAPSHOT_NAME_KEY);
+
+        // Clear the table region boundary cache to make sure long running jobs stay up to date
+        byte[] tableNameBytes = queryPlan.getTableRef().getTable().getPhysicalName().getBytes();
```

Comment:
This code is just part of the general refactor that Alex did in his original PR, the code was already there. Only the functions were moved.

Reply:
I think the code to clear the table region boundary cache was added as part of PHOENIX-2599. Since the MR/spark/pig integrations does not support order by which requires rows from a scan to be in sorted order we can skip the region boundary check in BaseScannerRegionObserver.preScannerOpen and also clear the region boundary cache. 
(see the discussion on PHOENIX-2599).